### ISSUE TITLE
[Homegear-Max] Implement virtual device for Window Switch and Temperature

### ISSUE THREAD

**Issue by Flole998** [Homegear/Homegear#313]
As Homegear-Max doesn't seem to accept issues I'm opening it here:

As already mentioned in the thread https://forum.homegear.eu/t/auf-windowswitchreceiver-schreiben/2466/16 I am looking for a way to write to the window switch receiver. As I mentioned aswell this is supported by FHEM and the code isn't overly complicated, so this should be possible here aswell when using the cc1101.

**Comment by hfedcba** [comment]
Yes it is doable, the problem is just the time to do it. I'll be happy to assist anybody who wants to add the function to the MAX module.

**Comment by Flole998** [comment]
I'm willing to do it, I just don't know what would be the best way to do it: Seperate channel? And require a pair command before using it? And implement directly in the Max module?

If I know where I should "connect" my code to then I can go from there, just finding the best spot to connect is difficult at the moment.

**Comment by hfedcba** [comment]
Probably you need to pair the thermostat to a "virtual device" (with HomeMatic BidCoS that is the case). You can look at the code for HomeMatic BidCoS as a starting point: https://github.com/Homegear/Homegear-HomeMaticBidCoS/blob/master/src/HomeMaticCentral.cpp#L2607

The code for MAX! should be similar. For HomeMatic BidCoS we are pairing Homegear to the window channel. For MAX! you would need to sniff first, how a real window contact sensor is paired to the thermostat. This communication needs then to be mimicked.

Cheers,

Sathya

**Comment by hfedcba** [comment]
(or see how it is done in FHEM for example)

**Comment by Flole998** [comment]
Yes exactly, pairing with a virtual device is necessary in this case aswell. And then it's just a single message that has to be sent from that virtual device. We already have this (which is the same as in Homematic):
https://github.com/Homegear/Homegear-MAX/blob/a4cdf083f7ec0eb36cdcbfc577fbd0d21920b4c6/src/MAXCentral.cpp#L1330

So I can just add that pairing there (at least for the windowswitch receiver, for setting the valve position directly that would probably cause issues as the valve will never receive any command and would not operate normally anymore)?

How would I go about sending the actual command then? Would I do that in https://github.com/Homegear/Homegear-MAX/blob/a4cdf083f7ec0eb36cdcbfc577fbd0d21920b4c6/src/MAXPeer.cpp#L1413 ?

What would happen if someone paired an actual window contact? Would this cause any issues? I definitely want to avoid breaking a function for everyone ;)

**Comment by hfedcba** [comment]
Hi @Flole998,

> So I can just add that pairing there

Yes.

> Would I do that in https://github.com/Homegear/Homegear-MAX/blob/a4cdf083f7ec0eb36cdcbfc577fbd0d21920b4c6/src/MAXPeer.cpp#L1413 ?

Yes.

What would happen if someone paired an actual window contact? Would this cause any issues?

No, that shouldn't cause any issues as multiple window contacts can be paired to the device.

Cheers,

Sathya

**Comment by Flole998** [comment]
Hi @hfedcba,

it took me a few days but I have made the first few modifications and packed them into a PR (so you can see what I've got so far). I haven't tested them yet though and I am almost 100% sure they aren't quite right yet.

How do I set the Channel correctly to tell the Thermostat that this is a Window Contact and should be linked accordingly? Or is this unrelated to the Channel? I would guess its channel 4 as QEQxxxxxx:4 is a WINDOW_SWITCH_RECEIVER? What is the senderChannelIndex for? Is that what I am looking for?

Also is there any kind of automatic build system I could use for my test builds? Otherwise I would have to setup some kind of crosscompiling for ARM aswell, that would be pretty annoying.

Kind regards

Flole

**Comment by kjabl** [comment]
Hello @hfedcba ,

as mentioned in the [german homegear forum ](https://forum.homegear.eu/t/entwicklung-virtueller-raumtemperaturf%C3%BChler/3647) I would like to give it a try at implementing a virtual wall thermostat as a source for a measured temperature and when I'm already in the topic I could also implement the virtual device for a window switch. That is why I would like to bring this issue back to life.

I started with looking at what Flole already did and what you already implemented for Homematic bidcos. I reimplemented the code from flole's pull request and made it more similar to what I found in HomeMaticCentral. I hope that will help when it comes to implementing different virtual devices for Window Switches and Wall Thermostats. 
At the moment I'm working at a class for the virtual peer "Wall Thermostat". I took the HmCcTC in the BidCoS family as an example and I'm looking at the code fhem uses as an example for building the packet.

But now I'm wondering if I need all of these DutyCycleFunctions in the my virtual peer class as there are in the virtual HmCcTc? As I'm not familiar with how BidCoS or MAX work in general I don't know how to adopt that for MAX. So I kinda hope that it would be enough to implement a function like setMeasuredTemperature which builds the packet, hands it over to the queue and sends it. Similar to the way we send the paring packet in HomeMaticCentral or in future also in MAXCentral. Could you give me some advice at these questions please? Or any hint where I can find the information I'm looking for?

Kind regards
Kevin

**Comment by hfedcba** [comment]
Hi @kjabl,

> But now I'm wondering if I need all of these DutyCycleFunctions in the my virtual peer class as there are in the virtual HmCcTc?

No, you don't. MAX! uses something called "wake on radio" (WOR), which actually is not a good thing, but makes communication easier. You can send packets anytime. Just make sure, you don't send them too often (each device or communication interface can only send 35 packets per hour).

> So I kinda hope that it would be enough to implement a function like setMeasuredTemperature which builds the packet, hands it over to the queue and sends it.

Yes, that's enough.

If you have any more questions, please contact me by e-mail (s.laufer@homegear.email).

Cheers,

Sathya

**Comment by kjabl** [comment]
Hi @hfedcba,

Sorry for the late reply, I was on vacation and very busy afterwards.

The information really helps me. I'll go back to working on it as soon as possible.

Best Regards
Kevin

**Comment by fgather** [comment]
Hey @kjabl ,

if I can support you in your efforts, I'd be happy to test or write code, since I also desperately need virtual devices (window and wall thermostats) for my MAX3 set-up. 

Best Regards,
Florian

**Comment by andyboeh** [comment]
@kjabl, @fgather Same here, I'm also willing to code, but I have to finish a few other projects first.

**Comment by kjabl** [comment]
@fgather @andyboeh ,
Oh I'm really sorry. I was on holiday when I read your message and totally forgot to answer it when I got back home.
As I got all my hardware in productive mode (cause its cold at the moment :D) I didn't really work on it. I need to check my actual state, but any help would be appreciated. After a quick look the last thing I did was to enable creating a virtual Wall thermostat "BCTCCWM4". I also wrote a class that represents this virtual thermostat. After compiling I tried to load the MAX module into my running homegear but got the error message that BCTCCWM4 doesn't exist and for that the module couldn't be loaded.
I'll test this again as soon as possible (I hope I can manage tomorrow), with a new clean compiled environment.

But in the meantime, what would be the best way to share what I did so far? I'm not really familiar working with git. Should I create a pull request so my code is uploaded and you can access it on here? I'm humbling a bit with that because the actual state isn't executable.

**Comment by Flole998** [comment]
You should probably create a fork and then you can work on it and everyone can track your progress. Once you're happy with it you can generate a pull request from the fork so it can be squashed and merged.

**Comment by kjabl** [comment]
thank you for your advise @Flole998 I just created a fork and pushed my work. You can find it here:
[](https://github.com/kjabl/Homegear-MAX)

**Comment by kjabl** [comment]
Just a quick status update: I got my second CUL today so I could test. Creation of a virtual wall thermostat is working now. When pairing a new valve drive, the virtual wallthermostat is automatically created by homegear and paired to valve drive. You can also see the virtual device in homegear's device list.
Next step is to implement the packet creation which I already started. 
Im now thinking about if it is a good idea to automatically create the virtual device upon pairing or if it would be better to have the option for manually creating it, if someone doesn't want these virtual devices.

**Comment by hfedcba** [comment]
> Im now thinking about if it is a good idea to automatically create the virtual device upon pairing or if it would be better to have the option for manually creating it

I'd say, it should be created on pairing - that makes this feature easily usable.

**Comment by Flole998** [comment]
@kjabl I've seen that you've made quite some progress. Will you also add support for virtual window contacts or only virtual wall thermostats?

**Comment by kjabl** [comment]
Hi @Flole998 I will also add a virtual window contact. For the moment it even looks like that will be easier. I already had a working setup last weekend. But after I few minutes problems started. That was because the valve drives want to "hear something from the wall thermostat every few minutes" (around 15 minutes as far as I know). Because of the auto-ack function of the culfw I switched to only one virtual wall thermostat for every valve drive. Perhaps I'll have to go back for individuals again.
For window Contacts I think I will definitely go for individual virtual devices instead of writing directly to the valve drive in homegear. Because at the moment there is no such variable in the device definition of the valve drive. But the valve drive doesn't contact the window switch which makes it a lot easier. When I find time to work on it next time I think I'll also implement that and test together.

**Comment by Flole998** [comment]
What problems started? Could you describe what happened then?

I think fhem doesn't send periodic stuff? They have this working, so we can copy from them ;)

**Comment by kjabl** [comment]
After some time my valve drive started sending wakeup packages (message type F1 if I remember correctly).

I also think fhem does send periodic, as in the fhem wiki they mention a Perl script which can be used to send periodic and should be used with fakewallthermostat. https://wiki.fhem.de/wiki/MAX
They also say that after 30 minutes the valve drive goes back to its internal thermometer. Which could be a way to handle this. I didn't have time to test this again. But the problem is that the wakeup packages are so frequent I'm afraid it will run out of credits before 30 minutes are over.
In the fhem forum they are also discussing this wakeup problem. I can look up the link later when I'm on my computer. Resending the fakewallthermostat packet then would be the best way I think, as it would prevent the valve drive from going back to its internal thermometer. Also the valve drive should be awake so we can send non burst to save credits.
But I'll have to test this again when I find time.

**Comment by Flole998** [comment]
Do you have an actual wall thermostat to check how it behaves?

**Comment by kjabl** [comment]
No sadly I don't which makes this more complicated.
But I think I read about it in fhem, don't remember exactly.
But in fact there are two options I think. An ack or a new packet. I think I can test this.

Also if we are using one fake address for all fakewallthermostats the CUL would be able to send auto acks. I configured this in my test but it didn't. So I guess an ack isn't the correct solution

**Comment by kjabl** [comment]
As promised i looked up the link at the fhem forum:
https://forum.fhem.de/index.php/topic,116493.0.html
They are discussing the problem I had with wakeup calls and it seems that it happens when the fake thermostat waits too long with sending. Can also lead to problems with the valve drives thinking a window was opened for example.

Another important link would be: https://forum.fhem.de/index.php/topic,115018.30.html
After some quick reading I think they found a solution so I will look into fhem again.

**Comment by mikeS7** [comment]
I can grab and post sniffed log from my real system. Will try on weekend or earlier.

**Comment by kjabl** [comment]
That would be great, thanks!

**Comment by mikeS7** [comment]
Parsed log from the room with 2 radiator thermostats - 30 mins
[GuestRoom-record.json.zip](https://github.com/Homegear/Homegear/files/6443906/GuestRoom-record.json.zip)

**Comment by mikeS7** [comment]
Log from the room with 1 thermostat
[Fireplace-record.json.zip](https://github.com/Homegear/Homegear/files/6443940/Fireplace-record.json.zip)

**Comment by mikeS7** [comment]
@kjabl wallthermostat should periodically send packets with type 0x42, check the `delay` fields in the logs
I need to update my parser to process shuttercontacts

**Comment by kjabl** [comment]
Great thanks @mikeS7 I think I'll have time tomorrow to look into it.
But that sounds good, the periodic 42 messages are used to send the actual measured and desired temperature.
F1 messages were a problem with my last test. But I found out more about them at the fhem forum and also wrote some code which I will have to test.

The window contact is kinda easy, I already tried building the hex string of the packet by hand and sending it manually. Works fine. The difficult part with the window contact is for me how to integrate it into homegear and how to make it writable. Pairing a virtual device and sending packets could be done really quick

**Comment by mikeS7** [comment]
@kjabl I can record logs with specific scenarios if required, but only on next 2 days

**Comment by kjabl** [comment]
@mikeS7 well if it is not too much trouble two scenarios would be interesting:
1 - pairing wall thermostat with valve drive
2 - the temperature measured by the wall thermostat suddenly falling down (that is what can cause wakeup packages and I'm not 100% sure if I found the right solution to that at fhem)

**Comment by kjabl** [comment]
@mikeS7 I tested some more today and found another scenario.

Can you record what happens if you go to valve drive and turn the wheel to change the target temperature? That makes the valve drive send something but I'm not sure what the correct answer would be

**Comment by mikeS7** [comment]
@kjabl I don't understand 2nd scenario. What is an issue with wakeup packages? As I remember correctly I saw wakeup packages only after restart or battery change, cube sends wakeup packages to shuttercontacts after restart.

I will record for room with 1 radiator thermostat 1 wall thermostat and 1 shuttercontact controlled by cube and standalone.

**Comment by kjabl** [comment]
ah ok, it's not about the normal wakeup packets. It's about 0xF1 in fhem they are called wakeup packet I don't really know if it is a WoR packet. If the wallthermostat sends a measured temperature far lower then the one before the valve drive kind of asks if its sure. (That's what I found out in fhem forum). I was able to send a packet by hand which stops the valve drive from going crazy, but I'm not sure if it is the right answer.
Today i observed a similar behaviour when setting the temperature manually at the valve drive while having it paired to a virtual wall thermostat.

**Comment by mikeS7** [comment]
Room with RadiatorThermostat, WallThermostat, WindowSwitch under Cube control.
Initial state: +16 Manual mode

1. RT +22
2. RT Auto mode (+23)
3. WT Eco (+16)
4. WT Manual mode
5. WT Auto mode (+23)
6. WT Manual mode
7. WT Eco (+16)
8. WT +18
9. RT Comfort (+23)
10. RT Auto mode (+23)
11. RT Manual Mode
12. RT Eco (+16)
13. WS Open (+12)
14. WS Close
[RT.WT.WS.Cube.json.zip](https://github.com/Homegear/Homegear/files/6445926/RT.WT.WS.Cube.json.zip)

**Comment by mikeS7** [comment]
Room with RadiatorThermostat, WallThermostat, WindowSwitch under Cube control.
Teach-out...

1. Delete WS with MAX! software -> system asks to activate the switch
2. Delete WT
3. Delete RT
4. Open the window
5. Close the window
[RT.WT.WS.Cube-TeachOut.json.zip](https://github.com/Homegear/Homegear/files/6446015/RT.WT.WS.Cube-TeachOut.json.zip)

**Comment by mikeS7** [comment]
@kjabl Could you please look for the pairing log, I can repeat if it's unclear

Room with RadiatorThermostat, WallThermostat, WindowSwitch  - no Cube

1. RT pairing mode, WT pairing mode -> success
2. RT pairing mode, WS pairing mode -> no success
3. RT pairing mode, WS pairing mode -> no success
4. WT pairing mode, WS pairing mode -> success
5. Window open
6. Window close
7. WT eco (+17)
[RT.WT.WS.Pairing.json.zip](https://github.com/Homegear/Homegear/files/6446052/RT.WT.WS.Pairing.json.zip)

**Comment by mikeS7** [comment]
Room with RadiatorThermostat, WallThermostat, WindowSwitch - no Cube
After unpairing WallThermostat RadiatorThermostat and WindowSwitch both go crazy with 0xf1 packets, they didn't receive ack for state packages sent to WT.
[RT.WT.WS.UnPairing.json.zip](https://github.com/Homegear/Homegear/files/6446418/RT.WT.WS.UnPairing.json.zip)

**Comment by mikeS7** [comment]
Pairing with Cube and setup
Room with RadiatorThermostat, WallThermostat, WindowSwitch
[RT.WT.WS.PairingCube.json.zip](https://github.com/Homegear/Homegear/files/6446427/RT.WT.WS.PairingCube.json.zip)

**Comment by Flole998** [comment]
@kjabl Were you able to get anything useful from those dumps or is there anything else that could help you?

**Comment by Flole998** [comment]
@kjabl During the last few days I've spent quite some time switching my Homegear setup to a single instance and the cc1101 being connected over homegear-gateway. Now I should be able to do some actual development and look into your work. Could you please give me a quick overview of the current status and what you need help with?

**Comment by kjabl** [comment]
Hi @Flole998 sorry for the late reply. I didn't have any time for this the last months.
I just pushed the actual state. 

The point I have problems with is stepping into homegears process of handling packets in the right situation. The code that I just pushed should work in theory. But when I tried testing it last time I couldn't pair my device anymore. I guess that was due to a timeout but I'm not sure, perhaps my CUL was a problem or anything else. I didn't find the time to look deeper into it.

The main difficulty is that you should have one virtual device in homegear for every fake device. Also FHEM went away from using one single address for all fake thermostats and one address for all fake shutter contacts. But with that you can't use the CUL's auto-ack feature anymore so homegear needs to do that (which I think isn't a problem in most of the cases).
But with the thermostats there's this F1 message type. Sometimes the valve drives start asking the fake wall thermostat for information about actual and target temperature. If you ignore this the valve drives will run out of credits very fast. So homegear needs to recognize this packet (addressed towards the fake wall thermostat & type F1) find the fitting virtual device build a packet with actual and target temperature and send the answer. This needs to happen as quick as possible because hopefully we can avoid sending with WoR then to save credits.
I (tried to) build this into homegear but afterwards I had problems pairing my valve drive so I couldn't test it.

**Comment by Flole998** [comment]
I just looked through your changes and noticed that it's kinda hard to follow. Could you please squash them together, then you can see everything you changed and I can comment directly in the commit to ask why you've done certain things in a certain way.

**Comment by kjabl** [comment]
I just sqashed them, hope it's easier to understand now

**Comment by Flole998** [comment]
I left a few comments, hopefully they help you finding the pairing issue

**Comment by Flole998** [comment]
@kjabl I've tested your changes and for me there's no issue pairing. I've seen that you've done some additional changes recently, are you still seeing that issue or is it working for you now aswell?

**Comment by Flole998** [comment]
@kjabl Is it still broken for you or could your work be merged now?

**Comment by aberridg** [comment]
Hi Guys,

Any progress on the PR or the issue? This is something I would very much like to use myself!

**Comment by Flole998** [comment]
@aberridg Unfortunately I still haven't checked what @kjabl has managed to do. It would probably be a good idea to check what's the current state of his repo and if it works or what issues there are.

**Comment by kjabl** [comment]
Sorry for my late reply, I really didnt have any time to work at it the last one and a half year. It is still broken I guess.
It works well i nothing "special" happens. But if for example the temperature falls down the valve drive starts asking the wallthermostat for the actual temperature. I still didn't figure out how to properly answer this. I know how this packet needs to look and I can build it. BUT I didn't find a way to access the actual measured temperature inside the code. Should be something around reading a parameter but I didn't get it right. It's here in the code: https://github.com/kjabl/Homegear-MAX/blob/master/src/MAXCentral.cpp#L253
And here https://github.com/kjabl/Homegear-MAX/blob/master/src/VirtualPeers/BcTcCWm.cpp#L120

Then there still stays the problem that answering all these Wakeups sent from valve drive to wallthermostat can consume a lot of credits and to save some credits we need to be really fast because then we can avoid waking up the valve drive again. Because of the problems I couldn't test the speed until today.
At the moment I am thinking of another way to do this. I could build a hardware device using an arduino or ESP8266/ESP32 microcontroller connect to a dedicated CUL device. This way we could avoid a lack of credits AND using a dedicated microcontroller gets us more to realtime communication then a linux system can be which again means saving credits. This could be attach to homegear by a rest call through node blue for example. But it's just an idea in hope to get around my problems. A solution inside homegear would be my favourite.

**Comment by Flole998** [comment]
There is an artificial delay introduced when sending the response normally. You can probably skip that and then you should be fast enough.

**Comment by Flole998** [comment]
Isn't the measured temperature stored in the `_newMeasuredTemperature` variable of the virtual peer?

**Comment by kjabl** [comment]
> Isn't the measured temperature stored in the `_newMeasuredTemperature` variable of the virtual peer?

That's a good point, yes it should be. We'll it's only stored there if it was set, so it's the last measured not necessarily the actual temperature but that would be fine enough.
For building the packet I also need the desired temperature. Which should be taken from exactly that device value to make sure it is the actual target temperature.

**Comment by kjabl** [comment]
The difficulty about the target temperature is where to find it. At the moment it is saved in the valve drive. 
My problem the last time I worked on it was not the "MAX part" but the homegear code and how to find the necessary data inside the code, access value from other homegear devices and so on.
Second problem is that I test with my live system/hardware and got into problems resetting and repairing them.

**Comment by Flole998** [comment]
I'm sure @hfedcba can help to get a certain parameter from a peer, in this case the set temperature from a certain valve.

Did you check what went wrong during resetting/pairing?

**Comment by kjabl** [comment]
> Did you check what went wrong during resetting/pairing?

I couldn't figure that out. I had problems even receiving the packets and then the packet came and homegear didn't react or the device reacted only once.
I think that was an hardware issue. The MAX devices are probably not designed to be reset 5 times a day

**Comment by Flole998** [comment]
That's possible, cause I tried your changes and didn't have issues pairing.
